physical/raft/raft.go

```diff
@@ -652,23 +619,140 @@ func (b *RaftBackend) DisableUpgradeMigration() (bool, bool) {
+		return
+	}
+
+	go func() {
+		ticker := time.NewTicker(b.verificationInterval())
+		defer ticker.Stop()
+
+		logger := b.logger.Named("raft-wal-verifier")
+
+		for {
+			select {
+			case <-ticker.C:
+				err := b.applyVerifierCheckpoint()
+				if err != nil {
+					logger.Error("error applying verification checkpoint", "error", err)
+				}
+				logger.Debug("sent verification checkpoint")
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (b *RaftBackend) applyVerifierCheckpoint() error {
+	data := make([]byte, 1)
+	data[0] = byte(verifierCheckpointOp)
+
+	b.permitPool.Acquire()
+	b.l.RLock()
+
+	var err error
+	applyFuture := b.raft.Apply(data, 0)
+	if e := applyFuture.Error(); e != nil {
+		err = e
+	}
+
+	b.l.RUnlock()
+	b.permitPool.Release()
+
+	return err
+}
+
+// isLogVerifyCheckpoint is the verifier.IsCheckpointFn that can decode our raft logs for
+// their type.
+func isLogVerifyCheckpoint(l *raft.Log) (bool, error) {
+	return isRaftLogVerifyCheckpoint(l), nil
+}
+
+func makeLogVerifyReportFn(logger log.Logger) verifier.ReportFn {
+	return func(r verifier.VerificationReport) {
+		if r.SkippedRange != nil {
+			logger.Warn("verification skipped range, consider decreasing validation interval if this is frequent",
+				"rangeStart", int64(r.SkippedRange.Start),
+				"rangeEnd", int64(r.SkippedRange.End),
+			)
+		}
+
+		l2 := logger.With(
+			"rangeStart", int64(r.Range.Start),
+			"rangeEnd", int64(r.Range.End),
+			"leaderChecksum", fmt.Sprintf("%08x", r.ExpectedSum),
+			"elapsed", r.Elapsed,
+		)
+
+		if r.Err == nil {
+			l2.Info("verification checksum OK",
+				"readChecksum", fmt.Sprintf("%08x", r.ReadSum),
+			)
+			return
+		}
+
+		if errors.Is(r.Err, verifier.ErrRangeMismatch) {
+			l2.Warn("verification checksum skipped as we don't have all logs in range")
+			return
+		}
+
+		var csErr verifier.ErrChecksumMismatch
+		if errors.As(r.Err, &csErr) {
+			if r.WrittenSum > 0 && r.WrittenSum != r.ExpectedSum {
+				// The failure occurred before the follower wrote to the log so it
+				// must be corrupted in flight from the leader!
+				l2.Error("verification checksum FAILED: in-flight corruption",
+					"followerWriteChecksum", fmt.Sprintf("%08x", r.WrittenSum),
+					"readChecksum", fmt.Sprintf("%08x", r.ReadSum),
+				)
+			} else {
+				l2.Error("verification checksum FAILED: storage corruption",
+					"followerWriteChecksum", fmt.Sprintf("%08x", r.WrittenSum),
+					"readChecksum", fmt.Sprintf("%08x", r.ReadSum),
+				)
+			}
+			return
+		}
+
+		// Some other unknown error occurred
+		l2.Error(r.Err.Error())
+	}
+}
+
 func (b *RaftBackend) CollectMetrics(sink *metricsutil.ClusterMetricSink) {
 	var stats map[string]string
+	var logStoreStats *etcdbolt.Stats
+
 	b.l.RLock()
-	logstoreStats := b.stableStore.(*raftboltdb.BoltStore).Stats()
-	fsmStats := b.fsm.Stats()
+	if boltStore, ok := b.stableStore.(*raftboltdb.BoltStore); ok {
+		bss := boltStore.Stats()
+		logStoreStats = &bss
+	}
+
```

Comment:
I wonder if we have tests that cover these metrics being produced? I don't see any usages of `CollectMetrics` in raft_test.go at least. If not it's probably worth (at least) adding to a TODO list for manual testing to be sure we don't regress those here.

Reply:
I'm not sure I'm following what kind of test you're looking for. `CollectMetrics` is called [here](https://github.com/hashicorp/vault/blob/main/vault/core_metrics.go#L144) as part of the main metrics collection loop. Can you be more specific?

Reply:
We talked offline but for the sake of GH history, I just meant that I don't think we have unit tests that assert that boltdb metrics are actually reported when this is called either locally or in general in Vault.

So it would be possible to typo this change and not fail any tests but break our actual metrics around boltdb.

Ideally we'd have some sort of unit test there but since there is no precedent we can plan to add one later, but it would be wise to at least manually verify that when using BoltDB on this branch we still get stats output as before!